lib/mandate/callbacks.rb

```diff
@@ -0,0 +1,89 @@
+module Mandate
+  module Callbacks
+    class AbortError < RuntimeError
+    end
+
+    class Results
+      attr_reader :result, :errors
+
+      def initialize
+        @succeeded = false
+        @errors = []
+      end
+
+      def succeeded!(result)
+        @result = result
+        @succeeded = true
+      end
+
+      def add_error(error)
+        errors << error
+      end
+
+      def succeeded?
+        !!succeeded
+      end
+
+      def success(&block)
+        return unless succeeded?
+        block.call(result)
+      end
+
+      def failure(&block)
+        return if succeeded?
+        block.call(errors)
+      end
```

Comment:
Returning `self` on both of these would allow chaining:

``` ruby
summer.
  success { ... }.
  failure { ... }
```

It would mean you can't get the results of the block, but then it's difficult to say when you'd need that. If you need the successful results, you can get `.results`, you can get `.errors`.

A thing you might consider is naming them `on_success` and `on_failure`/`on_error`, because just `success` and `failure` read intuitively like properties that hold the result (successful or failed) rather than methods that expect callbacks.

Reply:
@AndrewRadev I'm now returning self. Thanks.

I'm still pondering the callback nomenclature.